Simulation/Environment/config.txt
BAUD_SERVOS=115200
PORT_SERVOS=SERVOS_PIPE
_PORT_SERVOS=COM7

BAUD_LIDAR=115200
PORT_LIDAR=LIDAR_PIPE
_PORT_LIDAR=COM9

TIP_TO_WRIST=140
WRIST_TO_KNEE=56
KNEE_TO_SHOULDER=28
CENTER_TO_SHOULDER=166
KNEE_TO_BOTTOM=50

KNEE_MIN_ANGLE=35
KNEE_MAX_ANGLE=135
WRIST_MIN_ANGLE=20
WRIST_MAX_ANGLE=100
SHOULDER_MIN_ANGLE=-90
SHOULDER_MAX_ANGLE=90

STANCE_KNEE=0.8
STANCE_WRIST=0.4
STANCE_SHOULDER=0.5

LEFT_FRONT_WRIST_CHANNEL=0
LEFT_FRONT_WRIST_MIN=900
LEFT_FRONT_WRIST_MAX=1700
LEFT_FRONT_KNEE_CHANNEL=1
LEFT_FRONT_KNEE_MIN=2000
LEFT_FRONT_KNEE_MAX=1000
LEFT_FRONT_SHOULDER_CHANNEL=2
LEFT_FRONT_SHOULDER_MIN=550
LEFT_FRONT_SHOULDER_MAX=2370
LEFT_FRONT_DIRECTION=120

LEFT_MID_WRIST_CHANNEL=6
LEFT_MID_WRIST_MIN=800
LEFT_MID_WRIST_MAX=1600
LEFT_MID_KNEE_CHANNEL=5
LEFT_MID_KNEE_MIN=2200
LEFT_MID_KNEE_MAX=1100
LEFT_MID_SHOULDER_CHANNEL=4
LEFT_MID_SHOULDER_MIN=550
LEFT_MID_SHOULDER_MAX=2370
LEFT_MID_DIRECTION=180

LEFT_REAR_WRIST_CHANNEL=9
LEFT_REAR_WRIST_MIN=900
LEFT_REAR_WRIST_MAX=1650
LEFT_REAR_KNEE_CHANNEL=10
LEFT_REAR_KNEE_MIN=2100
LEFT_REAR_KNEE_MAX=1100
LEFT_REAR_SHOULDER_CHANNEL=8
LEFT_REAR_SHOULDER_MIN=550
LEFT_REAR_SHOULDER_MAX=2370
LEFT_REAR_DIRECTION=240

RIGHT_FRONT_WRIST_CHANNEL=17
RIGHT_FRONT_WRIST_MIN=2100
RIGHT_FRONT_WRIST_MAX=1300
RIGHT_FRONT_KNEE_CHANNEL=18
RIGHT_FRONT_KNEE_MIN=950
RIGHT_FRONT_KNEE_MAX=2000
RIGHT_FRONT_SHOULDER_CHANNEL=16
RIGHT_FRONT_SHOULDER_MIN=500
RIGHT_FRONT_SHOULDER_MAX=2300
RIGHT_FRONT_DIRECTION=60

RIGHT_MID_WRIST_CHANNEL=21
RIGHT_MID_WRIST_MIN=2250
RIGHT_MID_WRIST_MAX=1400
RIGHT_MID_KNEE_CHANNEL=22
RIGHT_MID_KNEE_MIN=900
RIGHT_MID_KNEE_MAX=1950
RIGHT_MID_SHOULDER_CHANNEL=20
RIGHT_MID_SHOULDER_MIN=550
RIGHT_MID_SHOULDER_MAX=2400
RIGHT_MID_DIRECTION=0

RIGHT_REAR_WRIST_CHANNEL=25
RIGHT_REAR_WRIST_MIN=2200
RIGHT_REAR_WRIST_MAX=1350
RIGHT_REAR_KNEE_CHANNEL=26
RIGHT_REAR_KNEE_MIN=850
RIGHT_REAR_KNEE_MAX=1900
RIGHT_REAR_SHOULDER_CHANNEL=24
RIGHT_REAR_SHOULDER_MIN=550
RIGHT_REAR_SHOULDER_MAX=2370
RIGHT_REAR_DIRECTION=300


Simulation/Hexapod/cmake-build-debug/config.txt
BAUD_SERVOS=115200
PORT_SERVOS=SERVOS_PIPE
_PORT_SERVOS=COM6

BAUD_LIDAR=115200
PORT_LIDAR=LIDAR_PIPE
_PORT_LIDAR=COM8

TIP_TO_WRIST=140
WRIST_TO_KNEE=56
KNEE_TO_SHOULDER=28
CENTER_TO_SHOULDER=166
KNEE_TO_BOTTOM=50

KNEE_MIN_ANGLE=35
KNEE_MAX_ANGLE=135
WRIST_MIN_ANGLE=20
WRIST_MAX_ANGLE=100
SHOULDER_MIN_ANGLE=-90
SHOULDER_MAX_ANGLE=90

STANCE_KNEE=0.8
STANCE_WRIST=0.4
STANCE_SHOULDER=0.5

LEFT_FRONT_WRIST_CHANNEL=0
LEFT_FRONT_WRIST_MIN=900
LEFT_FRONT_WRIST_MAX=1700
LEFT_FRONT_KNEE_CHANNEL=1
LEFT_FRONT_KNEE_MIN=2000
LEFT_FRONT_KNEE_MAX=1000
LEFT_FRONT_SHOULDER_CHANNEL=2
LEFT_FRONT_SHOULDER_MIN=550
LEFT_FRONT_SHOULDER_MAX=2370
LEFT_FRONT_DIRECTION=120

LEFT_MID_WRIST_CHANNEL=6
LEFT_MID_WRIST_MIN=800
LEFT_MID_WRIST_MAX=1600
LEFT_MID_KNEE_CHANNEL=5
LEFT_MID_KNEE_MIN=2200
LEFT_MID_KNEE_MAX=1100
LEFT_MID_SHOULDER_CHANNEL=4
LEFT_MID_SHOULDER_MIN=550
LEFT_MID_SHOULDER_MAX=2370
LEFT_MID_DIRECTION=180

LEFT_REAR_WRIST_CHANNEL=9
LEFT_REAR_WRIST_MIN=900
LEFT_REAR_WRIST_MAX=1650
LEFT_REAR_KNEE_CHANNEL=10
LEFT_REAR_KNEE_MIN=2100
LEFT_REAR_KNEE_MAX=1100
LEFT_REAR_SHOULDER_CHANNEL=8
LEFT_REAR_SHOULDER_MIN=550
LEFT_REAR_SHOULDER_MAX=2370
LEFT_REAR_DIRECTION=240

RIGHT_FRONT_WRIST_CHANNEL=17
RIGHT_FRONT_WRIST_MIN=2100
RIGHT_FRONT_WRIST_MAX=1300
RIGHT_FRONT_KNEE_CHANNEL=18
RIGHT_FRONT_KNEE_MIN=950
RIGHT_FRONT_KNEE_MAX=2000
RIGHT_FRONT_SHOULDER_CHANNEL=16
RIGHT_FRONT_SHOULDER_MIN=500
RIGHT_FRONT_SHOULDER_MAX=2300
RIGHT_FRONT_DIRECTION=60

RIGHT_MID_WRIST_CHANNEL=21
RIGHT_MID_WRIST_MIN=2250
RIGHT_MID_WRIST_MAX=1400
RIGHT_MID_KNEE_CHANNEL=22
RIGHT_MID_KNEE_MIN=900
RIGHT_MID_KNEE_MAX=1950
RIGHT_MID_SHOULDER_CHANNEL=20
RIGHT_MID_SHOULDER_MIN=550
RIGHT_MID_SHOULDER_MAX=2400
RIGHT_MID_DIRECTION=0

RIGHT_REAR_WRIST_CHANNEL=25
RIGHT_REAR_WRIST_MIN=2200
RIGHT_REAR_WRIST_MAX=1350
RIGHT_REAR_KNEE_CHANNEL=26
RIGHT_REAR_KNEE_MIN=850
RIGHT_REAR_KNEE_MAX=1900
RIGHT_REAR_SHOULDER_CHANNEL=24
RIGHT_REAR_SHOULDER_MIN=550
RIGHT_REAR_SHOULDER_MAX=2370
RIGHT_REAR_DIRECTION=300


Simulation/Hexapod/Config.cpp
#include "Config.h"
#include <map>
#include <string>
#include <iostream>
#include <tuple>
#include <sstream>
#include "Serial.h"
#include "Hexapod.h"
#include <cstdint>
#define _USE_MATH_DEFINES
#include <cmath>
#include <exception>

using namespace std;

map<string, string> Config::dictionary;

void Config::loadConfiguration()
{
    string port = Config::getValue<string>("PORT_SERVOS");
    auto speed = Config::getValue<long>("BAUD_SERVOS");
    speed_t baud = 0;
#ifdef _WIN32
    baud = static_cast<speed_t>(speed);
#else
    switch (speed) {
        case 2400: baud = B2400; break;
        case 38400: baud = B38400; break;
        case 9600: baud = B9600; break;
        case 115200: baud = B115200; break;
        default: throw runtime_error("Invalid BAUD in config");
    }
#endif // _WIN32
    servosPort.setup(port, baud);

    port = Config::getValue<string>("PORT_LIDAR");
    speed = Config::getValue<long>("BAUD_LIDAR");
    baud = 0;
#ifdef _WIN32
    baud = static_cast<speed_t>(speed);
#else
    switch (speed) {
        case 2400: baud = B2400; break;
        case 38400: baud = B38400; break;
        case 9600: baud = B9600; break;
        case 115200: baud = B115200; break;
        default: throw runtime_error("Invalid BAUD in config");
    }
#endif // _WIN32
    lidarPort.setup(port, baud);

    Geometry g {};
    g.tipToWrist = Config::getValue<float>("TIP_TO_WRIST");
    g.wristToKnee = Config::getValue<float>("WRIST_TO_KNEE");
    g.kneeToShoulder = Config::getValue<float>("KNEE_TO_SHOULDER");
    g.centerToShoulder = Config::getValue<float>("CENTER_TO_SHOULDER");
    g.kneeToBottom = Config::getValue<float>("KNEE_TO_BOTTOM");
    g.kneeMinAngle = Config::getValue<float>("KNEE_MIN_ANGLE")*M_PI/180.0;
    g.kneeMaxAngle = Config::getValue<float>("KNEE_MAX_ANGLE")*M_PI/180.0;
    g.wristMinAngle = Config::getValue<float>("WRIST_MIN_ANGLE")*M_PI/180.0;
    g.wristMaxAngle = Config::getValue<float>("WRIST_MAX_ANGLE")*M_PI/180.0;
    g.shoulderMinAngle = Config::getValue<float>("SHOULDER_MIN_ANGLE")*M_PI/180.0;
    g.shoulderMaxAngle = Config::getValue<float>("SHOULDER_MAX_ANGLE")*M_PI/180.0;
    g.stanceKneeNormalized = Config::getValue<float>("STANCE_KNEE");
    g.stanceWristNormalized = Config::getValue<float>("STANCE_WRIST");
    g.stanceShoulderNormalized = Config::getValue<float>("STANCE_SHOULDER");
    hexapod.setGeometry(g);

    Config::loadLimbConfig(liLeftFront, "LEFT_FRONT");
    Config::loadLimbConfig(liLeftMiddle, "LEFT_MID");
    Config::loadLimbConfig(liLeftRear, "LEFT_REAR");
    Config::loadLimbConfig(liRightFront, "RIGHT_FRONT");
    Config::loadLimbConfig(liRightMiddle, "RIGHT_MID");
    Config::loadLimbConfig(liRightRear, "RIGHT_REAR");
}

void Config::loadLimbConfig(LimbIndex index, std::string prefix)
{
    hexapod.initLimb(index,
        hexapod.initJoint(Config::getValue<int>(prefix+"_WRIST_CHANNEL"),
                          Config::getValue<float>(prefix+"_WRIST_MIN"),
                          Config::getValue<float>(prefix+"_WRIST_MAX")),
        hexapod.initJoint(Config::getValue<int>(prefix+"_KNEE_CHANNEL"),
                          Config::getValue<float>(prefix+"_KNEE_MIN"),
                          Config::getValue<float>(prefix+"_KNEE_MAX")),
        hexapod.initJoint(Config::getValue<int>(prefix+"_SHOULDER_CHANNEL"),
                          Config::getValue<float>(prefix+"_SHOULDER_MIN"),
                          Config::getValue<float>(prefix+"_SHOULDER_MAX")),
        Config::getValue<float>(prefix+"_DIRECTION")*M_PI/180
    );
}

bool Config::load()
{
    FILE *configFile = fopen("config.txt", "r");
    if (!configFile) {
        cerr << "Couldn't load config";
        return false;
    }

    char line[200];
    while (fgets(line, 200, configFile)) {
        tuple<string, string> item;
        if (Config::splitKeyValue(line, item)) {
            dictionary[get<0>(item)] = get<1>(item);
        }
    }

    try {
        Config::loadConfiguration();
    } catch (exception &e) {
        cerr << e.what() << "\n";
        return false;
    }

    fclose(configFile);
    return true;
}

bool Config::splitKeyValue(char *line, std::tuple<std::string, std::string> &keyAndValue)
{
    int i = 0;
    while (line[i]) {
        if (line[i] == '=') {
            line[i] = 0;
            string key(line), value(line+i+1);
            keyAndValue = make_tuple(key, value);
            return true;
        }
        ++i;
    }
    return false;
}

template <typename T> T Config::getValue(std::string key)
{
    auto i = dictionary.find(key);
    if (i == dictionary.end()) {
        throw runtime_error("Config key " + key + " not found");
    }
    string value = dictionary[key];
    istringstream stream(value);
    T converted;
    stream >> converted;
    return converted;
}
